crates/typst-kit/src/package.rs

```diff
@@ -14,6 +15,111 @@ use typst::syntax::package::{
+            path: lock_file_path,
+            tmp_path: format!("{parent_dir_path}/{lock_file_name}"),
+        })
+    }
+
+    /// Try acquiring exclusive lock.
+    fn try_lock(&self) -> io::Result<()> {
+        self.file.try_lock_exclusive()
+    }
+
+    /// Acquire exclusive lock.
+    fn lock(&self) -> Result<(), PackageError> {
+        self.file
+            .lock_exclusive()
+            .map_err(other_err("failed to aquire lock file"))
+    }
+
+    /// Try acquiring the exclusive lock, if failed print an error an wait
+    /// for the lock with blocking.
+    ///
+    /// Returns `Ok(true)` if acquired first try, otherwise `Ok(false)`.
+    fn try_and_lock(&self) -> Result<bool, PackageError> {
+        match self.try_lock() {
+            Ok(_) => {
+                eprintln!("[tmp] acquired the lock file {:?}", self.tmp_path);
+                Ok(true)
+            }
+            Err(ref err) if err.kind() == io::ErrorKind::WouldBlock => {
+                eprintln!("[tmp] couldn't aquire the lock file {:?}", self.tmp_path);
+                eprintln!(
+                    "Waiting for another instance to finish installing the package..."
+                );
+                self.lock()?;
+
+                // // If other instance successfully installed a package after
+                // // waiting for the lock, then there is nothing left to do.
+                // if dir.exists() {
+                //     eprintln!("[tmp] dropped the lock file {:?}", self.tmp_path);
+                //     return Ok(dir);
+                // }
+
+                // eprintln!(
+                //     "Another instance failed to install the package, trying it again."
+                // );
+                Ok(false)
+                // Additional cool/user-friendly logs:
+                // let _write_lock = lock_file.write().unwrap();
+                // println!("checking if package already exists");
+                // if dir.exists() {
+                //     println!("it does! no need to download and unpack");
+                //     return Ok(dir);
+                // } else {
+                //     println!("waited to aquire the file lock and the other instance wasn't able to download it :/");
+                //     todo!()
+                // }
+            }
+            Err(err) => Err(other_err("failed to aquire lock file")(err)),
+        }
+    }
+
+    /// Remove (delete) lock file (if wasn't already by other process).
+    ///
+    /// Will be automatically called when struct is being dropped ([Drop]).
+    fn remove(&self) -> Result<(), PackageError> {
+        if self.path.exists() {
+            fs::remove_file(&self.path)
+                .map_err(other_err("failed to remove lock file"))?;
+        }
+        eprintln!("[tmp] dropped the lock file {:?}", self.tmp_path);
+        Ok(())
+    }
+}
+
+impl Drop for LockFile {
+    fn drop(&mut self) {
+        self.remove().expect("An error occured when dropping a lock file")
```

Comment:
This may be called during a panic, a panic within a panic causes an immediate abort, and I'm not sure if we want that. Perhaps a simple print to stderr is better (through `writeln!(io::stderr(), ...)`, since `eprintln!` would panic for a closed stream, I believe). Including a file path would probably be important too.

Reply:
> This may be called during a panic, a panic within a panic causes an immediate abort, and I'm not sure if we want that.

Pretty sure 1 or 2 panics wouldn't make any difference, other than show backtrace by default.



> Perhaps a simple print to stderr is better

Yeah.

> since `eprintln!` would panic for a closed stream, I believe

No, for a normal panic it works normally.

Reply:
> Pretty sure 1 or 2 panics wouldn't make any difference, other than show backtrace by default.

No, a panic within a panic aborts unconditionally, that's bad for many reasons, one of them being the fact that the user won't get any useful error message.

> No, for a normal panic it works normally.

What is a normal panic exactly? In any case, I did some reading, and it turns out `eprintln` doesn't panic when writing to `stderr`, while `println` will. This is completely unrelated to the current panic state, though.